cmd/maya-apiserver/spc-actions/select_disk.go

```diff
@@ -0,0 +1,207 @@
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	openebs "github.com/openebs/maya/pkg/client/generated/clientset/internalclientset"
+	"github.com/openebs/maya/pkg/client/k8s"
+)
+
+// clientset struct holds the interface of internalclientset
+// i.e. openebs.
+// This struct will be binded to method ListDisk and is useful in mocking
+// and unit testing.
+type clientSet struct {
+	oecs openebs.Interface
+}
+
+// nodeDisk struct will be used as a value for a map nodeDiskMap (map defined in ListDisk function)
+// The struct will be useful in forming the data structure nodeDiskMap which will be manipulated
+// to efficiently select the nodes and disk for dynamic pool provisioning.
+type nodeDisk struct {
+	// diskCount is the count of usable disks that can be used in storagepool provisioning.
+	diskCount int
+	//diskList is the list of usable disks that can be used in storagepool provisioning.
+	diskList []string
+}
+
+// spareAllotment holds the value for the remaining node allotments
+// for pool provisioning.
+var spareAllotment int16
+
+// getDiskList is a wrapper function which will receive list of disks
+// that can be used for dynamic pool provisioning at runtime.
+// The function finally returns the disk list to the caller (i.e. getCasPoolDisk function).
+func getDiskList(cp *v1alpha1.CasPool) ([]string, error) {
+
+	// Get kubernetes clientset
+	// namespaces is not required, hence passed empty.
+	newK8sClient, err := k8s.NewK8sClient("")
+
+	if err != nil {
+		return nil, err
+	}
+	// Get openebs clientset using a getter method (i.e. GetOECS() ) as
+	// the openebs clientset is not exported.
+	newOecsClient := newK8sClient.GetOECS()
+
+	// Create instance of clientset struct defined above which binds
+	// ListDisk method and fill it with openebs clienset (i.e.newOecsClient ).
+	newClientSet := clientSet{
+		oecs: newOecsClient,
+	}
+
+	// if no minimum pools were specified it will default to 1.
+	if cp.MinPools <= 0 {
+		glog.Warning("invalid or 0 min pool specified, defaulting to 1")
+		cp.MinPools = 1
+	}
+
+	// nodeDiskAlloter will try to return a list of disks so that maxpool number of storagepool
+	// is provisioned.
+	diskList, err := newClientSet.nodeDiskAlloter(cp)
+	if err != nil {
+		return nil, err
+	}
+
+	return diskList, nil
+}
+
+// nodeDiskAlloter will try to allot nodes for pool creation as specified in
+// maxPool field of the storagepoolclaim.
+
+// For exapmle, if maxPool=5 and minPool=3, it will try to search for 5 nodes that will qualify for
+// pool provisioning. At least 3 node should qualify else pool will not be provisioned and pool creation
+// will be aborted gracefully with proper log messages.
+
+// If no minPool field is present,at least one node must qalify for pool provisioning.
+
+// modeDiskAlloter can be made more intelligent as per the required pool constraints for alloting nodes.
+func (k *clientSet) nodeDiskAlloter(cp *v1alpha1.CasPool) ([]string, error) {
```

Comment:
Trying to think in terms of struct design:

- These are the first things that come to my mind
```go
type poolSelector interface {
  Select() (nodes []string, error)
}

type poolSelectorSpec struct {
  // properties that are required while running the pool selection strategies
}

type poolSelectorStrategy func(spec poolSelectorSpec) (nodes []string, err error)

// selectPool implements poolSelector interface
type selectPool struct {
  spec poolSelectorSpec
  poolSelectors  []poolSelectorStrategy
}

func SelectPool(...) *selectPool {
  return &selectPool{...}
}

func openebsSelector(spec poolSelectorSpec) (nodes []string, err error) {}
func nodeSelector(spec poolSelectorSpec) (nodes []string, err error) {}
func zoneSelector(spec poolSelectorSpec) (nodes []string, err error) {}

```

Reply:
There is no idea of pool selector.
Just select disks and nodes and form pool.
Finally only disks imformation matter for pool creation not the node.
We discuss it in detail.
But I am trying to understand your struct design .

Reply:
What matters for users? I was trying to think from user's perspective. 
Should we abstract the disks from users?
If yes, how do we? What is that thing that user understands better?
Perhaps a set of `poolCreatorStrategies`.
You may think on the names.

Reply:
Ok.... we might need to discuss it.
Many a things will come out after thatm
Anyways, let us consider these as an enhancement which will go as part of next PR?